### DISCUSSION TITLE
Addressing SQL Injection Vulnerabilities with Raw Methods.

### DISCUSSION THREAD

**Discussion by craigfrancis** [laravel/framework#47257] [Ideas]
While the [documentation](https://laravel.com/docs/10.x/queries#raw-methods) notes "Laravel can not guarantee that any query using raw expressions is protected against SQL injection vulnerabilities", unfortunately not everyone will read/understand/remember this; and even the best developers make mistakes.

While Injection Vulnerabilities are becoming rarer (thanks to database abstractions like yours), they have not been solved, and these mistakes are still found in penetration tests (as reflected in the OWASP Top 10). They also tend to go un-reported, as it's kinda embarrassing, especially for developers who think they are more "experienced" and believe they should use the "powerful" methods for everything (sigh).

A simple example - I found someone had used `$sort = $request->input('sort')` with `->orderBy($sort)`, all good, fairly normal. But a junior developer had been asked to split the `name` into first/last fields, so `$sort` was set to `name_first,name_last`; this caused an SQL error (good), but instead of updating their code to call `->orderBy()` for each field, they simply changed it to `->orderByRaw($sort)`, it "worked", and was uploaded to the live site.

And then you have these (abbreviated) classics, would you say these mistakes are obvious?

```php
$users = DB::table('user')->whereRaw('CONCAT(name_first, " ", name_last) LIKE "' . $search . '%"'); // INSECURE
$users = DB::table('user')->whereRaw('CONCAT(name_first, " ", name_last) LIKE ?', $search . '%');

$users = DB::select('SELECT * FROM users WHERE active = ' . $_GET['active']); // INSECURE
$users = DB::select('SELECT * FROM users WHERE active = ?', [$_GET['active']]);
```

I believe Laravel should provide a way to check unsafe user input is not used with these sensitive parameters... i.e. these parameters should expect a "developer defined string", known as the `literal-string` type in Static Analysis tools [PHPStan](https://github.com/phpstan/phpstan/releases/tag/0.12.97) and [Psalm](https://github.com/vimeo/psalm/releases/tag/4.8.0) (also supported by [PhpStorm  2022.3](https://youtrack.jetbrains.com/issue/WI-64109/literal-string-support-in-phpdoc)).

This is used by [Nette](https://github.com/nette/database/commit/fb2476b2f7937053a99d30b53c7e5731f6f7b96c), [Propel](https://github.com/propelorm/Propel2/pull/1788/files), [RedBean](https://github.com/gabordemooij/redbean/pull/873/files); and it does [find mistakes](https://twitter.com/OndrejMirtes/status/1525481771348934657).

As an aside, [Pradeep Srinivasan](https://us.pycon.org/2022/speaker/profile/23/) and [Graham Bleaney](https://bleaney.ca/) (Facebook/Meta) introduced the [LiteralString type](https://peps.python.org/pep-0675/) in Python 3.11; and this technique can be used in [other programming languages](https://eiv.dev/).

---

I can put together a PR, but want to check that you (the Laravel developers) want to address this issue, and are happy with this approach, e.g.

```diff
--- a/src/Illuminate/Database/Query/Builder.php
+++ b/src/Illuminate/Database/Query/Builder.php
@@ -1024,7 +1024,7
     /**
      * Add a raw where clause to the query.
      *
-     * @param  string  $sql
+     * @param  literal-string  $sql
      * @param  mixed  $bindings
      * @param  string  $boolean
      * @return $this
      */
     public function whereRaw($sql, $bindings = [], $boolean = 'and')
@@ -2353,7 +2353,7
     /**
      * Add a raw "order by" clause to the query.
      *
-     * @param  string  $sql
+     * @param  literal-string  $sql
      * @param  array  $bindings
      * @return $this
      */
     public function orderByRaw($sql, $bindings = [])
```

I'd prefer to start with a few of the simpler methods (to get some feedback), before we look at methods like `DB::raw()`, and `DB::select()`.

**Comment by Niki-Crawley** [comment]
Huh, that would solve some issues - last week I found `selectRaw()` being mis-used (don't ask).

Have you found any issues with this approach? maybe false-positives?

**Reply by craigfrancis** [reply]
Hi @Niki-Crawley; yeah, unfortunately these mistakes happen quite often.

Good question about issues, oddly I've not really found any so far... I've used this technique on 5 Laravel projects (via stubs), and it found a few more vulnerabilities I'd missed during my manual code reviews; and it highlighted 1 bit code that was less than ideal, where a developer had used `strtoupper()` for some unknown reason.

**Comment by tpetry** [comment]
That's a good approach! It will make potential SQL injection vulnerabilities a lot less likely if teams use PHPStan to analyze their code. And with my different PRs for query expressions, we can now even get rid of those raw expressions entirely and replace them with safe escaped generated sql constructs.

**Reply by craigfrancis** [reply]
Yep, agreed. While it would be nice for people to not use the Raw methods at all (i.e. using something safer, like your [Query Expressions](https://github.com/tpetry/laravel-query-expressions)), the `literal-string` type works with how code exists today, and identifies mistakes that lead to Injection Vulnerabilities.
